pkg/canary/deployer.go

```diff
@@ -339,6 +348,37 @@ func (c *Deployer) getSelectorLabel(deployment *appsv1.Deployment) (string, erro
 	return "", fmt.Errorf("selector not found")
 }
 
+var sidecars = map[string]bool{
+	"istio-proxy": true,
+	"envoy":       true,
+}
+
+// getPorts returns a list of all container ports
+func (c *Deployer) getPorts(deployment *appsv1.Deployment, canaryPort int32) (map[string]int32, error) {
+	ports := make(map[string]int32)
+
+	for _, container := range deployment.Spec.Template.Spec.Containers {
+		// exclude service mesh proxies based on container name
+		if _, ok := sidecars[container.Name]; ok {
+			continue
+		}
+		for i, p := range container.Ports {
+			// exclude canary.service.port
+			if p.ContainerPort == canaryPort {
+				continue
+			}
+			name := fmt.Sprintf("tcp-%v", i)
+			if p.Name != "" {
+				name = p.Name
+			}
+
+			ports[name] = p.ContainerPort
+		}
+	}
+
+	return ports, nil
```

Comment:
Does this take into consideration the following possible deployment?
```yaml
    containers:
         - name: my-grpc
            ports:
                - containerPort: 8000  
         - name: my-rest-proxy
            ports:
                - containerPort: 8080
```
Im not thoroughly familiar with go, if two containers have 1 port definition each, would one overwrite the other in the `ports` map?

Reply:
Another possible example showing different containers with duplicate port names
```yaml
    containers:
         - name: my-grpc
            ports:
                - containerPort: 8000  
                - containerPort: 9899
                   name: prom-metrics
                   protocol: TCP
         - name: my-rest-proxy
            ports:
                - containerPort: 8080
                - containerPort: 9899
                   name: prom-metrics
                   protocol: TCP
```

Reply:
You are right, I've appended the container name to avoid conflicts.